File: day3/puzzle1.clj
(require '[clojure.string :as str])

(defn get-coords [file]
  (vec (map #(str/split % #",") (str/split (slurp file) #"\n") )))

(defn get-manhattan-distance [p q]
  ; Find the manhattan distance between two points
  (let [[px py] p
        [qx qy] q
        distance (+ (Math/abs (- qx px)) (Math/abs (- qy py)))]
    distance))

(defn get-orientation [p q r]
  ; Find orientation of ordered triplet
  (let [[px py] p
        [qx qy] q
        [rx ry] r
        orientation (- (* (- qy py) (- rx qx)) (* (- qx px) (- ry qy)))]
    (cond 
      (= orientation 0) 0
      (> orientation 0) 1
      (< orientation 0) 2)))

(defn on-segment? [p q r]
  ;Does point p lie on line segment pr?
  (let [[px py] p
        [qx qy] q
        [rx ry] r]
    (cond
      (and 
       (<= qx (max px rx))
       (>= qx (min px rx))
       (<= qy (max py ry))
       (>= qy (min py ry))) true
      :else false)))

(defn calculate-intersection-point [p q r s]
  ; Calculate the intersection point
  ; http://www.ambrsoft.com/MathCalc/Line/TwoLinesIntersection/TwoLinesIntersection.htm
  (println p q r s)
  (let [[x1 y1] p
        [x2 y2] q
        [x3 y3] r 
        [x4 y4] s
        denom (- (* (- x2 x1) (- y4 y3)) (* (- x4 x3) (- y2 y1)))
        x (/ 
           (- (* (- (* x2 y1) (* x1 y2)) (- x4 x3)) (* (- (* x4 y3) (* x3 y4)) (- x2 x1))) denom)
        y (/ 
           (- (* (- (* x2 y1) (* x1 y2)) (- y4 y3)) (* (- (* x4 y3) (* x3 y4)) (- y2 y1))) denom)]
    [x y]))

(defn intersect? [pq pq2]
  ; Do these two lines intersect?
  ;(println pq pq2)
  (let [
        [p q] pq
        [p2 q2] pq2
        orientation-1 (get-orientation p q p2) 
        orientation-2 (get-orientation p q q2) 
        orientation-3 (get-orientation p2 q2 p) 
        orientation-4 (get-orientation p2 q2 q)]
    (cond
      ; General case
      (and (not= orientation-1 orientation-2) (not= orientation-3 orientation-4)) [pq pq2]
      ; Special cases
      (and (= orientation-1 0) (on-segment? p p2 q)) [pq pq2]
      (and (= orientation-2 0) (on-segment? p q2 q)) [pq pq2]
      (and (= orientation-3 0) (on-segment? p2 p q2)) [pq pq2]
      (and (= orientation-4 0) (on-segment? p2 q q2)) [pq pq2]
      :else nil)))
    

(defn get-intersection-points [line-segments]
  ; Return a list of intersection points
  (let [x (get line-segments 0) y  (get line-segments 1) cmp (compare (count x) (count y)) intersection-points []]
    (cond 
      (empty? x) intersection-points
      (= cmp 0) (for [i (range (count x)) j (range (count x))] (conj intersection-points 
                                                           (intersect? 
                                                            (nth x i) 
                                                            (nth y j))))
      (= cmp 1) (for [i (range (count x)) j (range (count y))] (conj intersection-points 
                                                                     (intersect? 
                                                                      (nth x i) 
                                                                      (nth y j)))))))

(defn get-points [source movement line]
  ; Get points for line
  ;(println line)
  (if (empty? movement) (conj line source)
      (let [
            [_ direction steps] (re-matches #"([RLUD])(\d+)" (first movement))
            num-steps (Integer/parseInt steps)]
        (cond
          (= direction "R") (get-points
                             (assoc source 0 (+ (get source 0) num-steps)) (drop 1 movement) (conj line source))
          (= direction "U") (get-points
                             (assoc source 1 (+ (get source 1) num-steps)) (drop 1 movement) (conj line source))
          (= direction "L") (get-points
                             (assoc source 0 (- (get source 0) num-steps)) (drop 1 movement) (conj line source))
          (= direction "D") (get-points
                             (assoc source 1 (- (get source 1) num-steps)) (drop 1 movement) (conj line source))))))

(defn get-line-segments [lines]
  ; break up the lines into segments
  ;(println lines)
  (let [segments [(partition 2 1 (get lines 0)) (partition 2 1 (get lines 1))]] segments))

(defn get-lines [coords]
  ; Get lines based on coords provided in input
  (loop [i 0 lines []]
    (if (= i 2)
      lines
      (recur (inc i) (conj lines (get-points [0 0] (get coords i) []))))))

(defn find-closest [intersection-points]
  ;Find closet intersection point to origin; We don't care about points that include [0 0]
  ;(println intersection-points)
  (let [filtered-points (remove (fn [x] (or (nil? (first x)) (= (nth (get (get x 0) 0) 0) [0 0]))) intersection-points)]
    (println filtered-points)
    (loop [i 0
           cnt (count filtered-points)
           distances []]
      (if (= i cnt) 
        distances
        (recur 
         (inc i) cnt 
         (conj distances 
               (get-manhattan-distance [0 0] (calculate-intersection-point (nth (get (get (nth filtered-points i) 0) 0) 0)
                                                                           (nth (get (get (nth filtered-points i) 0) 0) 1)
                                                                           (nth (get (get (nth filtered-points i) 0) 1) 0)
                                                                           (nth (get (get (nth filtered-points i) 0) 1) 1)))))))))

(println (apply min (find-closest (get-intersection-points (get-line-segments (get-lines (get-coords "input")))))))

File: day3/puzzle1.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	x int
	y int
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Get the various line segments from the input data
func parseInputLine(lineCoords []string) []Point {
	var line []Point
	line = append(line, Point{0, 0})
	count := 0
	for _, v := range lineCoords {
		coord, _ := strconv.Atoi(v[1:])
		switch direction := v[0]; direction {
		case 'R':
			line = append(line, Point{line[count].x + coord, line[count].y})
		case 'L':
			line = append(line, Point{line[count].x - coord, line[count].y})
		case 'U':
			line = append(line, Point{line[count].x, line[count].y + coord})
		case 'D':
			line = append(line, Point{line[count].x, line[count].y - coord})
		}
		count++
	}

	return line
}

//Find the manhattan distance between two points
func getManhattanDistance(x Point, y Point) int {
	return abs(y.x-x.x) + abs(y.y-x.y)
}

// Find orientation of ordered triplet
func getOrientation(p Point, q Point, r Point) int {
	var orientation = (q.y-p.y)*(r.x-q.x) - (q.x-p.x)*(r.y-q.y)

	if orientation == 0 {
		return 0
	} else if orientation > 0 {
		return 1
	} else {
		return 2
	}
}

// Does point p lie on line segment pr?
func isOnSegment(p Point, q Point, r Point) bool {
	if q.x <= max(p.x, r.x) && q.x >= min(p.x, r.x) &&
		q.y <= max(p.y, r.y) && q.y >= min(p.y, r.y) {
		return true
	}
	return false
}

// Do these two line segments intersect?
func findIntersectingSegments(p Point, q Point, p2 Point, q2 Point) []Point {
	var orientation1 = getOrientation(p, q, p2)
	var orientation2 = getOrientation(p, q, q2)
	var orientation3 = getOrientation(p2, q2, p)
	var orientation4 = getOrientation(p2, q2, q)
	intersectingSegments := []Point{p, q, p2, q2}

	// General Case
	if orientation1 != orientation2 && orientation3 != orientation4 {
		return intersectingSegments
	}

	// Special Cases
	if (orientation1 == 0 && isOnSegment(p, p2, q)) ||
		(orientation2 == 0 && isOnSegment(p, q2, q)) ||
		(orientation3 == 0 && isOnSegment(p2, p, q2)) ||
		(orientation4 == 0 && isOnSegment(p2, q, q2)) {
		return intersectingSegments
	}
	return nil
}

// Calculate the intersection point
// http://www.ambrsoft.com/MathCalc/Line/TwoLinesIntersection/TwoLinesIntersection.htm
func calculateIntersectionPoint(p Point, q Point, p2 Point, q2 Point) (Point, error) {
	denom := (q.x-p.x)*(q2.y-p2.y) - (q2.x-p2.x)*(q.y-p.y)
	if denom == 0 {
		return Point{}, errors.New("Nope")
	}
	x := ((q.x*p.y-p.x*q.y)*(q2.x-p2.x) - (q2.x*p2.y-p2.x*q2.y)*(q.x-p.x)) / denom
	y := ((q.x*p.y-p.x*q.y)*(q2.y-p2.y) - (q2.x*p2.y-p2.x*q2.y)*(q.y-p.y)) / denom
	return Point{x, y}, nil
}

// Find the smallest distance from a list of Manhattan distances
func findMinimumDistance(distances []int) int {
	min := distances[0]
	for _, v := range distances {
		if v < min {
			min = v
		}
	}

	return min
}

func calculateDistances(line1 []Point, line2 []Point, i int, j int) int {
	result := findIntersectingSegments(line1[i], line1[i+1], line2[j], line2[j+1])
	if result != nil {
		res, err := calculateIntersectionPoint(result[0], result[1], result[2], result[3])
		if err != nil {
			fmt.Println("Whoops", err)
			return 0
		}
		return getManhattanDistance(Point{0, 0}, res)
	}
	return 0
}

// Find the Manhattan distance of the intersecting points from the source
func findDistances(lines [][]Point) []int {
	line1 := lines[0]
	line2 := lines[1]

	line1Size := len(lines[0])
	line2Size := len(lines[1])

	var distances []int

	if line1Size == line2Size {
		for i := 0; i <= line1Size-2; i++ {
			for j := 0; j <= line1Size-2; j++ {
				distance := calculateDistances(line1, line2, i, j)
				if distance != 0 {
					distances = append(distances, distance)
				}
			}
		}
	} else if line1Size > line2Size {
		for i := 0; i <= line1Size-2; i++ {
			for j := 0; j <= line2Size-2; j++ {
				distance := calculateDistances(line1, line2, i, j)
				if distance != 0 {
					distances = append(distances, distance)
				}
			}
		}
	} else {
		for i := 0; i < line2Size-2; i++ {
			for j := 0; j < line1Size-2; j++ {
				distance := calculateDistances(line1, line2, i, j)
				if distance != 0 {
					distances = append(distances, distance)
				}
			}
		}
	}

	return distances
}

func main() {
	file, err := os.Open("input")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines [][]Point

	for scanner.Scan() {
		lines = append(lines, parseInputLine(strings.Split(scanner.Text(), ",")))
	}

	distances := findDistances(lines)
	fmt.Println(findMinimumDistance(distances))

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

}
